wikitexthtml/render/preprocess.py

```diff
@@ -73,6 +85,44 @@ def hr_line(instance: WikiTextHtml, body: str) -> str:
     return regex.sub(r"\n----+\n", r"\n\n<hr />\n", body)
 
 
+def code_block(instance: WikiTextHtml, body: str) -> str:
+    # Use Pygments to generate highlighted code blocks
+
+    # The SYNTAXHIGHLIGHT_PATTERN Generates 6 groups:
+    #   The first group (match[0]) is the entire match
+    #   The third group (match[2]) is the language
+    #   The fourth group (match[3]) is a boolean to specify line numbering
+    #   The fifth group (match[4]) is the line number to start at
+    #   The sixth group (match[5]) is the code to highlight
+    # You can also reference the groups by name, i.e. match.group("lang")
+    # An interactive example with test strings is at https://regex101.com/r/9KjJYa/1
+
+    for match in reversed(list(SYNTAXHIGHLIGHT_PATTERN.finditer(body))):
+        lang = match.group("lang") if match.group("lang") else "unknown"
+        lineBoolean = True if match.group("line") else False
+        lineStart = int(match.group("lineStart")) if match.group("lineStart") else 1
+        content = match.group("code")
+
+        try:
+            # If language was provided, use that lexer
+            lexer = get_lexer_by_name(lang)
```

Comment:
Should you already go in this case when the lang is `unknown`? Maybe skip this step in that case? (as I suspect most people won't bother setting a language :P)

Reply:
I'm trying to follow the "[EAFP](https://devblogs.microsoft.com/python/idiomatic-python-eafp-versus-lbyl/)" python convention. What you suggest is certainly a valid choice, and is in-use for C-family langs.

I'm trying to express in code that the guessing lexer is really the less-preferred (read: much less performant) choice than specified lang -- regardless of what the user does.

Reply:
Yeah, ironically, I am talking neither about EAFP nor about LBYL ;)

My problem is with your default value `unknown`. Nowhere in your code you express what your expectations are with it, besides the implicit assumption pygments doesn't have a highlighter that handles `unknown`. But if it would know `unknown` as language, your code would do something completely different than your intentions. So I think that is a code-smell, and needs some kind of resolution :)

I see two options:

1) instead of doing the `if/else` on line 101 (`lang = ..`), expand the `if/else` to call `guess_lexer` directly. That is a clean way of expressing yourself.
2) use `None` as fallback value (instead of `unknown`). https://github.com/pygments/pygments/blob/master/pygments/lexers/__init__.py#L104 tells me that is a safe value to always abort from `get_lexer_by_name`. However, if you go that route, it does require a comment to state your intention here :)

Explicit before implicit, when it comes to fallback scenarios. Make it known to the reader what you expecting the code to do .. future-us will greatly thank you :D

Reply:
Ah, I see. I'll give that a revision!

Reply:
@TrueBrain how does this look? https://gist.github.com/Qrbaker/11ed01ae22434adb86990cca99deddf5

Reply:
I think you now went for both options :P I am totally fine you want to do EAFP, which works lovely for when users enter weird language names we don't support. So I am fine if you leave the value `None` and do what you did :)

Reply:
PS: `match` already returns None if it cannot find the group; `or None` doesn't add value :D Sorry for the nitpick here ;)

Reply:
You caught it between revisions, I mangled the paste.

I'm just trying to assign lang, and if it doesn't get assigned, fall back to guessing (match.group("group name") returns none if the group exists but is not captured)

Reply:
Hihi :)

Right, so this is option 1) I guess. Now I see it, I don't like it. Mainly, it no longer captures the exception, in cases you use a language that doesn't exist in pygments. And I am internally heavily debating what I think of that ... ideally that would return an error to the user, saying: yo, I don't support this! But .. there is no infrastructure for that. So honestly .. I kinda fancy'd what you did before, and just fall back to auto-detect, to make something useful out of it.

So personally, I would go option 2 route, I think. Something like:

```python
        lang = match.group("lang")

        try:
            # Try to use the language provided. If no language was giving, this will error out in a similar way as when the language is unknown to pygments.
            lexer = get_lexer_by_name(lang)
        except ClassNotFound:
            # Either there was no language, or it was unknown to pygments. Select the lexer best matching the code.
            lexer = guess_lexer(content)
```

Don't particularly fancy my comments, but you get what I mean :) Feel free to reword it ;)

Reply:
PS: this is called [bikeshedding](https://fourweekmba.com/bikeshedding/) :D It is all fine, really, I just want the bikeshed to be slightly more blue-ish :D

Reply:
> But .. there is no infrastructure for that.

Isn't there, kinda? How do we present the wikitext validation errors in TrueWiki? Would it be possible to hook into that object and add an invalid language warning?

Reply:
You are right!

`instance.add_error(f'{e.args[0][:-1]} (wikilink "{wikilink.string}").')`

From another file! Now that would be even better .. tell the user when he did something stupid .. and autodetect when we have no other way :D

Reply:
I will try and get that implemented in the next day or two.

Reply:
@TrueBrain added some error reporting & handling. See what you think:
![error_example_1](https://user-images.githubusercontent.com/14336407/163416192-3d52e209-4995-48a5-b331-dc9b532bedae.png)

Reply:
I think it is the best thing to do, report this error. That way we strongly hint to users to set the language, which is speedy. So that is nice :) A good addition to your work!